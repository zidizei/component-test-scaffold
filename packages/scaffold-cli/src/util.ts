import baseGlob from "glob"
import path from "path"

export async function glob(patterns: Array<string>): Promise<string[]> {
    const pattern = (patterns.length > 0) ? `+(${patterns.join("|")}).*` : "*"

    return new Promise((resolve, reject) => {
        baseGlob(path.resolve(`./**/__scaffolds__/${pattern}.scaffold.js`), (err, matches) => {
            if (err) { return reject(err) }
            resolve(matches)
        })
    })
}
