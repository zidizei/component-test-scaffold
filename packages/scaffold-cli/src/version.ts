const {version: VERSION} = require("../package.json")

export default (): string => VERSION
