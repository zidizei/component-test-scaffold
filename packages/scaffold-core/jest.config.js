const baseConfig = require("../../jest.config");
const package = require("./package.json");

module.exports = Object.assign({}, baseConfig, {
    rootDir: ".",
    displayName: package.name,
    setupFiles: [
        "./test/setup.ts"
    ],
    globals: {
        "ts-jest": {
            tsConfig: "<rootDir>/test/tsconfig.json"
        }
    }
});
