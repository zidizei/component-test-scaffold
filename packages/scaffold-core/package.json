{
  "name": "component-test-scaffold",
  "version": "0.0.2-alpha.0",
  "description": "Core package for scaffolding templates from a backend server for testing.",
  "main": "dist/index.js",
  "scripts": {
    "build": "tsc --build .",
    "clean": "rimraf dist build",
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "author": "Patrick Lam <zidizei@gmail.com>",
  "license": "MIT",
  "dependencies": {
    "jsdom": "^15.1.0",
    "node-fetch": "^2.6.0",
    "prettier": "^1.17.1"
  }
}
