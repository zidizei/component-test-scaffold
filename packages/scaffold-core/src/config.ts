import urlUtil from "url"

import { ScaffoldOptions } from "."

const getFilenameFromModule = (parent: NodeModule | null) => parent && parent.filename

const getCasenameFromUrl = (url: string) => {
    const pathname = urlUtil.parse(url).pathname

    return (pathname === undefined || pathname === "/") ? url : pathname
}

export function configure(url: string, opts: Partial<ScaffoldOptions>, parent: NodeModule | null): ScaffoldOptions {

    const filename = opts.filename || getFilenameFromModule(parent)
    const casename = opts.casename || getCasenameFromUrl(url)

    if (filename === null) {
        throw new Error("Cannot determine Test Suite file name. Try specifiying it manually using the 'filename' option.")
    }

    return Object.assign({}, opts, { filename, casename, fallback: true })
}
