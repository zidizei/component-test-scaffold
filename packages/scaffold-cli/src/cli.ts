import yargs from "yargs"

import { update } from "./index"
import version from "./version"

const CLI_VERSION = version()
const CLI_USAGE = "Usage: $0 [patterns..]"
const CLI_OPTS = {
    verbose: {
        demandOption: false,
        describe: "Show more details for updated scaffold files",
        type: 'boolean' as 'boolean',
    }
}

export type Parsed = {
    [x: string]: unknown
    verbose: boolean | undefined
    $0: string
}

export async function run(argv?: Array<string>) {
    argv = argv || process.argv.slice(2)

    const {_: patterns, ...parsed} = yargs(argv)
        .usage(CLI_USAGE)
        .options(CLI_OPTS)
        .version(CLI_VERSION)
        .alias("h", "help")
        .argv

    update(patterns, parsed)
}
