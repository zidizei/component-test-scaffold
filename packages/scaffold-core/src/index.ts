import https from "https"
import path from "path"

import { JSDOM } from "jsdom"

import { configure } from "./config"
import { loadUsingStrategy } from "./strategy"

export interface IScaffoldData {
    url: string
    test: string
    template: string
}

export interface IScaffoldResult {
    url: string
    template: string
    dom: JSDOM
}

export type Scaffold = {
    totalScaffolds: number
    currentScaffold: IScaffoldData
}

export type ScaffoldOptions = {
    filename: string
    casename: string
    writeback?: boolean
    loadStrategy?: "networkFirst" | "networkOnly" | "cacheFirst" | "cacheOnly" // TODO
    agent?: https.AgentOptions
}

export default async function (url: string, opts = {} as Partial<ScaffoldOptions>): Promise<IScaffoldResult> {
    const options = configure(url, opts, module.parent)

    const scaffoldPath = path.join(path.dirname(options.filename), "__scaffolds__")
    const scaffoldFile = `${path.basename(options.filename, path.extname(options.filename))}.scaffold.js`
    const scaffoldLocation = path.join(scaffoldPath, scaffoldFile)

    const html = await loadUsingStrategy(url, scaffoldLocation, options)
    const {template} = html.currentScaffold

    return {
        url,
        template,
        dom: new JSDOM(template, { url }),
    }
}
