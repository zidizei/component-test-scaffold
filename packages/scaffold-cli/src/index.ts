import load, { IScaffoldData } from "@component-test/scaffold-core"

import { Parsed } from "./cli"
import { glob } from "./util"

type Summary = {
    files: number
}

export async function update(patterns: Array<string>, {verbose}: Parsed): Promise<Summary> {
    const matches = await glob(patterns)
    const summary = {
        files: matches.length
    }

    await matches.forEach(async (filePath) => {
        const scaffold: { [c: string]: IScaffoldData } = require(filePath)
        const updates = Object.keys(scaffold).map((casename) => {
            const cas = scaffold[casename]

            if (verbose) {
                console.log("Updating '" + casename + "'.")
            }

            return load(cas.url, {
                filename: cas.test,
                casename,
                writeback: true,
                loadStrategy: "networkOnly"
            })
        })

        await Promise.all(updates)
    })

    return summary
}
