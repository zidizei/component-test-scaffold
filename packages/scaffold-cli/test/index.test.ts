import baseGlob from "glob"
import path from "path"
import fs from "fs"

import { IScaffoldData } from "@component-test/scaffold-core"

import { log } from "../src/util"
import { Parsed } from "../src/cli"
import { update } from "../src"

describe("Scaffold CLI", () => {

    const filename = __filename
    const template = "<body><strong>Test</strong></body>"
    const template2 = "<body><strong>Test 2</strong></body>"

    const url = "https://localhost:8080/patterns/component"

    const scaffoldLocation = path.resolve(
        __dirname,
        "./__scaffolds__/index.test.scaffold.js"
    )

    jest.mock("./__scaffolds__/index.test.scaffold.js", () => ({}))

    const mockedScaffoldData = require("./__scaffolds__/index.test.scaffold.js")
    const mockedGlob = baseGlob as unknown as jest.Mock // Let TypeScript know that 'glob' has been mocked.

    let spyWriteFile: jest.SpyInstance
    let spyLogVerbose: jest.SpyInstance
    let spyLogInfo: jest.SpyInstance
    let spyLogError: jest.SpyInstance

    let _stdOut = ""
    let _stdError = ""

    beforeEach(() => {
        mockedGlob.mockRestore()
        fetchMock.resetMocks()
        spyWriteFile = jest.spyOn(fs, "writeFile")
        spyLogVerbose = jest.spyOn(log, "verbose")
        spyLogInfo = jest.spyOn(log, "info")
        spyLogError = jest.spyOn(log, "error")
        _stdOut = ""
        _stdError = ""
    })

    afterEach(() => {
        jest.restoreAllMocks()
    })

    it("can update multiple Scaffold files", async () => {
        fetchMock.mockResponse(template)

        spyLogVerbose.mockImplementation((msg) => { _stdOut += msg + "\n" })
        spyLogInfo.mockImplementation((msg) => { _stdOut += msg + "\n" })
        spyWriteFile.mockImplementation((_filePath, _data, callback) => {
            callback(null)
        })

        mockedGlob.mockImplementationOnce((_path, callback) => {
            callback(null, [scaffoldLocation, scaffoldLocation])
        })

        const casename = "test case"
        const casedata: IScaffoldData = {
            url,
            test: path.relative(path.dirname(scaffoldLocation), filename),
            template
        }

        mockedScaffoldData[casename] = casedata

        const actual = await update([], { verbose: true } as Parsed)
        expect(actual.files).toBe(2)
        expect(spyLogVerbose).toBeCalledTimes(2)
        expect(spyLogInfo).toBeCalledTimes(1)
        expect(_stdOut).toMatchSnapshot()
    })

    it("can update single Scaffold file", async () => {
        fetchMock.mockResponse(template2)

        spyLogVerbose.mockImplementation((msg) => { _stdOut += msg + "\n" })
        spyLogInfo.mockImplementation((msg) => { _stdOut += msg + "\n" })
        spyWriteFile.mockImplementation((_filePath, _data, callback) => {
            callback(null)
        })

        mockedGlob.mockImplementationOnce((_path, callback) => {
            callback(null, [scaffoldLocation])
        })

        const casename = "test case"
        const casedata: IScaffoldData = {
            url,
            test: path.relative(path.dirname(scaffoldLocation), filename),
            template
        }

        mockedScaffoldData[casename] = casedata

        const actual = await update([], { verbose: false } as Parsed)
        expect(actual.hasError).toBeFalsy()
        expect(actual.files).toBe(1)
        expect(spyLogInfo).toBeCalledTimes(1)
        expect(_stdOut).toMatchSnapshot()
    })

    it("can update pattern-matched Scaffold file", async () => {
        fetchMock.mockResponse(template2)

        spyLogVerbose.mockImplementation((msg) => { _stdOut += msg + "\n" })
        spyLogInfo.mockImplementation((msg) => { _stdOut += msg + "\n" })
        spyWriteFile.mockImplementation((_filePath, _data, callback) => {
            callback(null)
        })

        mockedGlob.mockImplementationOnce((_path, callback) => {
            callback(null, [scaffoldLocation])
        })

        const casename = "test case"
        const casedata: IScaffoldData = {
            url,
            test: path.relative(path.dirname(scaffoldLocation), filename),
            template
        }

        mockedScaffoldData[casename] = casedata

        const actual = await update(["pattern.test.js"], { verbose: false } as Parsed)
        expect(actual.hasError).toBeFalsy()
        expect(actual.files).toBe(1)
        expect(spyLogInfo).toBeCalledTimes(2)
        expect(_stdOut).toMatchSnapshot()
    })

    it("displays a message when an Error is being thrown", async () => {
        fetchMock.mockResponseOnce(template, { status: 400 })

        spyLogInfo.mockImplementation((msg) => { _stdOut += msg + "\n" })
        spyLogError.mockImplementation((msg) => { _stdError += msg + "\n" })

        mockedGlob.mockImplementationOnce((_path, callback) => {
            callback(null, [scaffoldLocation])
        })

        const actual = await update([], {} as Parsed)
        expect(actual.hasError).toBeTruthy()
        expect(spyLogInfo).toBeCalledTimes(1)
        expect(spyLogError).toBeCalledTimes(1)
        expect(_stdError).toMatchSnapshot()
    })

})
