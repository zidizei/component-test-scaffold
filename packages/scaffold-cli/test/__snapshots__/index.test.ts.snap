// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scaffold CLI can update multiple Scaffold files 1`] = `
"Updating [1m2[22m Scaffold files.
Updating 'test case'.
Updating 'test case'.
"
`;

exports[`Scaffold CLI can update pattern-matched Scaffold file 1`] = `
"Updating [1m1[22m Scaffold file.
Matching patterns: [1mpattern.test.js[22m
"
`;

exports[`Scaffold CLI can update single Scaffold file 1`] = `
"Updating [1m1[22m Scaffold file.
"
`;

exports[`Scaffold CLI displays a message when an Error is being thrown 1`] = `
"LoadFromUrlError: Failed to load Scaffold from URL 'https://localhost:8080/patterns/component'.
"
`;
